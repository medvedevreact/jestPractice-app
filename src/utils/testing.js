import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

export const renderWithRouter = (component, options = {}) => {
  const Wrapper = ({ children }) => (
    <MemoryRouter {...options}>{children}</MemoryRouter>
  );
  return render(component, { wrapper: Wrapper });
};
