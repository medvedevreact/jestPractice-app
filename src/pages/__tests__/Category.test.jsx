import { renderWithRouter } from '../../utils/testing';
import * as api from '../../api';
import { waitForElementToBeRemoved, screen } from '@testing-library/react';
import { Category } from '../Category';
import { useNavigate } from 'react-router-dom';
import userEvent from '@testing-library/user-event';

const apiSpy = jest.spyOn(api, 'getFilteredCategory');

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn(),
}));

const useNavigateMock = jest.mocked(useNavigate);

describe('Category', () => {
  test('render', async () => {
    apiSpy.mockResolvedValueOnce({
      meals: [
        {
          strMeal: 'Beef and Mustard Pie',
          strMealThumb:
            'https://www.themealdb.com/images/media/meals/sytuqu1511553755.jpg',
          idMeal: '52874',
        },
        {
          strMeal: 'Beef and Oyster pie',
          strMealThumb:
            'https://www.themealdb.com/images/media/meals/wrssvt1511556563.jpg',
          idMeal: '52878',
        },
      ],
    });

    renderWithRouter(<Category />);
    const preloader = screen.getByRole('progressbar');
    expect(preloader).toBeInTheDocument();
    await waitForElementToBeRemoved(preloader);
    expect(screen.getAllByRole('article')).toMatchSnapshot();
  });
  test('navigates back when "Go Back" button is clicked', async () => {
    const navigateMock = jest.fn();

    useNavigateMock.mockReturnValue(navigateMock);

    apiSpy.mockResolvedValueOnce({
      meals: [
        {
          strMeal: 'Beef and Mustard Pie',
          strMealThumb:
            'https://www.themealdb.com/images/media/meals/sytuqu1511553755.jpg',
          idMeal: '52874',
        },
      ],
    });

    renderWithRouter(<Category />);

    const preloader = screen.getByRole('progressbar');
    await waitForElementToBeRemoved(preloader);

    const goBackButton = screen.getByRole('button', { name: /go back/i });

    await userEvent.click(goBackButton);

    expect(navigateMock).toHaveBeenCalledWith(-1);
    expect(navigateMock).toHaveBeenCalledTimes(1);
  });
});
