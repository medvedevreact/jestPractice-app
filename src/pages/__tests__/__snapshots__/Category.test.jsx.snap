// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Category render 1`] = `
[
  <div
    class="card"
    role="article"
  >
    <div
      class="card-image"
    >
      <img
        alt="Beef and Mustard Pie"
        src="https://www.themealdb.com/images/media/meals/sytuqu1511553755.jpg"
      />
    </div>
    <div
      class="card-content"
    >
      <span
        class="card-title"
      >
        Beef and Mustard Pie
      </span>
    </div>
    <div
      class="card-action"
    >
      <a
        class="btn"
        href="/meal/52874"
      >
        Watch recipe
      </a>
    </div>
  </div>,
  <div
    class="card"
    role="article"
  >
    <div
      class="card-image"
    >
      <img
        alt="Beef and Oyster pie"
        src="https://www.themealdb.com/images/media/meals/wrssvt1511556563.jpg"
      />
    </div>
    <div
      class="card-content"
    >
      <span
        class="card-title"
      >
        Beef and Oyster pie
      </span>
    </div>
    <div
      class="card-action"
    >
      <a
        class="btn"
        href="/meal/52878"
      >
        Watch recipe
      </a>
    </div>
  </div>,
]
`;
